File: src/coreclr/tests/src/Interop/COM/NativeClients/Primitives/ColorTests.cpp
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

#include "ClientTests.h"

#define RED RGB(0xFF, 0x00, 0x00)
#define GREEN RGB(0x00, 0xFF, 0x00)

namespace
{
    void VerifyColorMarshalling(IColorTesting* color)
    {
        HRESULT hr;
        BOOL match;

        THROW_IF_FAILED(color->AreColorsEqual(GREEN, GREEN, &match));

        THROW_FAIL_IF_FALSE(match);
    }

    void VerifyGetRed(IColorTesting* color)
    {
        HRESULT hr;
        OLE_COLOR red;
        
        THROW_IF_FAILED(color->GetRed(&red));
        
        THROW_FAIL_IF_FALSE(red == RED);
    }
}

void Run_ColorTests()
{
    HRESULT hr;

    CoreShimComActivation csact{ W("NETServer.dll"), W("ColorTesting") };

    ComSmartPtr<IColorTesting> color;
    THROW_IF_FAILED(::CoCreateInstance(CLSID_ColorTesting, nullptr, CLSCTX_INPROC, IID_IColorTesting, (void**)&color));

    VerifyColorMarshalling(color);
    VerifyGetRed(color);
}

File: src/clr/tests/src/Interop/COM/NETClients/Primitives/ColorTests.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

namespace NetClient
{
    using System;
    using System.Drawing;
    using System.Runtime.InteropServices;
    using TestLibrary;

    class ColorTests
    {
        private readonly Server.Contract.Servers.ColorTesting server;
        public ColorTests()
        {
            this.server = (Server.Contract.Servers.ColorTesting)new Server.Contract.Servers.ColorTestingClass();
        }

        public void Run()
        {
            this.VerifyColorMarshalling();
            this.VerifyGetRed();
        }

        private void VerifyColorMarshalling()
        {
            Assert.IsTrue(server.AreColorsEqual(Color.Green, ColorTranslator.ToOle(Color.Green)));
        }

        private void VerifyGetRed()
        {
            Assert.AreEqual(Color.Red, server.GetRed());
        }
    }
}
